File: apps/api/internal/agent/wordcount_test.go
package agent

import "testing"

func TestCountWords(t *testing.T) {
	cases := []struct {
		name string
		in   string
		want int
	}{
		{"empty", "", 0},
		{"whitespace only", "   \n\t ", 0},
		{"latin words", "the quick brown fox", 4},
		{"cjk chars each count", "中国是否让地球更可持续", 11},
		{"mixed cjk and latin", "中国的 GDP 增长", 6}, // 中 国 的 (3) + GDP (1) + 增 长 (2) = 6
		{"latin with punctuation", "hello, world!", 2},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			if got := CountWords(c.in); got != c.want {
				t.Errorf("CountWords(%q) = %d, want %d", c.in, got, c.want)
			}
		})
	}
}

File: apps/web/test/workspace/wordcount.test.ts
import { describe, it, expect } from "vitest";
import { countWords } from "@/workspace/blocks/wordcount";

// Locks in BUG-02 (2026-08-25 findings): the writing-room live counter must
// count words the same CJK-aware way the backend (agent.CountWords) does, so
// the "字" number agrees with the 整稿体检 review and the word-budget band. The
// old `text.replace(/\s+/g,"").length` counted characters — ~5× too high for
// English against an "约 800 词" target.
describe("countWords (mirrors backend agent.CountWords)", () => {
  it("counts English by words, not characters", () => {
    expect(countWords("China alone accounts for a quarter of the greening")).toBe(9);
  });

  it("counts CJK by character", () => {
    expect(countWords("中国让地球更可持续")).toBe(9);
  });

  it("mixes CJK chars and Latin words", () => {
    // 中 国 的 在 (4 Han) + "GDP" (1) + "2024" (1) = 6
    expect(countWords("中国的 GDP 在 2024")).toBe(6);
  });

  it("ignores extra whitespace and newlines", () => {
    expect(countWords("  hello   world \n\n foo ")).toBe(3);
  });

  it("is zero for empty / whitespace-only", () => {
    expect(countWords("")).toBe(0);
    expect(countWords("   \n\t ")).toBe(0);
  });
});
